### PR TITLE
Int 6291 ingest polymer account

### PR DESCRIPTION
**Pull request by adam-in-ict**
# Description

Thank you for contributing to a JupiterOne integration!

Please include a summary of the change and which issue is fixed. Please also
include relevant motivation and context. List any dependencies that are required
for this change.

## Summary

Initial implementation of Polymer integration.  More relationships and updates to come in later PRs.

image

## Type of change

Please leave any irrelevant options unchecked.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [ ] This change requires a documentation update

### General Development Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### Integration Development Checklist:

Please leave any irrelevant options unchecked.

- [X] I have checked for additional permissions required to call any new API
      endpoints, and have documented any additional permissions in
      `jupiterone.md`, where necessary.
- [X] My changes properly paginate the target service provider's API
- [X] My changes properly handle rate limiting of the target service provider's
      API
- [X] My new integration step is instrumented to execute in the correct order
      using `dependsOn`
- [X] I have referred to the
      [JupiterOne data model](https://github.com/JupiterOne/data-model/tree/main/src/schemas)
      to ensure that any new entities/relationships, and relevant properties,
      match the recommended model for this class of data
- [X] I have updated the `CHANGELOG.md` file to describe my changes
- [ ] When changes include modifications to existing graph data ingestion, I've
      reviewed all existing managed questions referencing the entities,
      relationships, and their property names, to ensure those questions still
      function with my changes.

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -1,2 +1,2 @@
-CLIENT_ID=
-CLIENT_SECRET=
\ No newline at end of file
+API_TOKEN=
+BASE_URL=
```
Comment (zemberdotnet) [inline .env.example:2]: I see your config has `organization`. Does this need that as well?

File: src/client.ts [hunk 2]
```diff
@@ -1,124 +1,187 @@
...
-        name: 'User Two',
-      },
-    ];
+    let queryUrl: string | null = this.requestUrl;
+    do {
+      this.logger.info({ queryUrl }, 'Calling API with url');
+      const response = await this.executeAPIRequestWithRetries<PolymerResponse>(
+        queryUrl,
+      );
+      // this.logger.info({ response }, `Full response`);
```
Comment (zemberdotnet) [inline src/client.ts:91]:
```suggestion
```

File: src/steps/account/index.test.ts [hunk 3]
```diff
@@ -13,6 +17,9 @@ test('fetch-account', async () => {
   recording = setupProjectRecording({
     directory: __dirname,
     name: 'fetch-account',
+    options: {
+      mode: isRecordingEnabled() ? 'record' : 'replay',
+    },
```
Comment (zemberdotnet) [inline src/steps/account/index.test.ts:22]: Would these be good contributions to make to the `integration-template`?

File: src/steps/violations/converter.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        name: violation.name,
+        createdOn: parseTimePropertyValue(violation.time),
+        category: 'other',
+        severity: 'unknown', // TODO (adam-in-ict) If we begin getting a category and severity, update these
+        numericSeverity: 0,
+        open: true,
+      },
+    },
+  });
+}
```
Comment (zemberdotnet) [inline src/steps/violations/converter.ts:65]: Are we going to include the `file` property from `PolymerItem` on another entity or should we add it to this one? Or not include it.

File: src/steps/violations/index.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  const accountEntity = (await jobState.getData(ACCOUNT_ENTITY_KEY)) as Entity;
+
+  await apiClient.iterateViolations(async (violation) => {
+    let violationEntity = await jobState.findEntity(
+      generateViolationKey(violation.id),
+    );
+    if (!violationEntity) {
+      violationEntity = await jobState.addEntity(
+        createViolationEntity(violation),
+      );
```
Comment (zemberdotnet) [inline src/steps/violations/index.ts:36]: Were you seeing duplicates during development?

File: src/steps/violations/index.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            ruleEntity,
+            violationEntity,
+            Number(rule.count),
+          ),
+        );
+      }
+    }
+    // else {
+    //   logger.error(violation, `Received duplicate violation entity key`)
+    // }
```
Comment (zemberdotnet) [inline src/steps/violations/index.ts:57]:
```suggestion
```
Delete if we don't need any logs. I forget if support custom metrics yet or not.

File: docs/jupiterone.md [hunk 7]
```diff
@@ -1,35 +1,23 @@
-# {{provider}}
+# Polymer Integration with JupiterOne
```
Comment (zemberdotnet) [inline docs/jupiterone.md:1]: These days we're just doing `# Polymer`

File: .env.example (hunk 1 above)
Comment (adam-in-ict) [reply .env.example:2]: Yep.  Thanks!  Missed that update when I changed the name.

File: src/steps/account/index.test.ts (hunk 3 above)
Comment (adam-in-ict) [reply src/steps/account/index.test.ts:22]: Possibly?  The main issue here is that the URL itself changes when we go from loading the .env to not loading the .env, so we have to make sure it doesn't try to record one that's going to fail.  Most other integrations won't run into this.

File: src/steps/violations/index.ts (hunk 5 above)
Comment (adam-in-ict) [reply src/steps/violations/index.ts:36]: Oops.  I was, but I'd like to remove that now.  We shouldn't see violation duplicates, but I'll confirm and update.  Thanks!

File: .env.example (hunk 1 above)
Comment (adam-in-ict) [reply .env.example]: Updated in latest commit.

File: src/steps/violations/index.ts (hunk 5 above)
Comment (adam-in-ict) [reply src/steps/violations/index.ts]: Updated in latest commit after testing again.  Thanks!

File: src/steps/violations/index.ts (hunk 6 above)
Comment (adam-in-ict) [reply src/steps/violations/index.ts]: Removed in latest commit along with duplicate check.  Thanks!

File: docs/jupiterone.md (hunk 7 above)
Comment (adam-in-ict) [reply docs/jupiterone.md]: Thanks!  Updated in latest commit.

File: src/steps/violations/converter.ts (hunk 4 above)
Comment (adam-in-ict) [reply src/steps/violations/converter.ts]: On second thought, I decided to go ahead and add that as preliminary info to the violations in this PR.  It should be out there in the latest commit.  Thanks!

File: src/steps/account/index.test.ts (hunk 3 above)
Comment (zemberdotnet) [reply src/steps/account/index.test.ts:22]: Ah gotcha.

**Review by zemberdotnet — COMMENTED** [review]
One more small change then it's good for a stamp.

**Review by zemberdotnet — APPROVED** [review]

**Pull request merged**